spec/controllers/users_controller_spec.rb

```diff
@@ -120,34 +120,78 @@
+          end
+
+          it_behaves_like 'tracks opt-in time', :opted_in_newsletter
+          it_behaves_like 'tracks opt-in time', :opted_in_announcements
+          it_behaves_like 'tracks opt-in time', :opted_in_marketing_announcements
+          it_behaves_like 'tracks opt-in time', :opted_in_surveys
+          it_behaves_like 'tracks opt-in time', :opted_in_sponsorships
+          it_behaves_like 'tracks opt-in time', :opted_in_applications_open
+
+          context 'opting out' do
+            shared_examples_for 'tracks opt-out' do |attribute|
+              before do
+                user.update_attribute attribute, Time.new(2002, 10, 31)
+              end
+
+              it "sets #{attribute} to nil" do
+                put :update, params: { id: user.to_param, user: { attribute => '0' } }
+                expect(user.reload.send "#{attribute}_at").to eq(nil)
+              end
+            end
+
+            it_behaves_like 'tracks opt-out', :opted_in_newsletter
+            it_behaves_like 'tracks opt-out', :opted_in_announcements
+            it_behaves_like 'tracks opt-out', :opted_in_marketing_announcements
+            it_behaves_like 'tracks opt-out', :opted_in_surveys
+            it_behaves_like 'tracks opt-out', :opted_in_sponsorships
+            it_behaves_like 'tracks opt-out', :opted_in_applications_open
+          end
         end
       end
+    end
 
-      context "another user's profile" do
-        let!(:another_user) { create(:user) }
+    describe "with invalid params" do
+      it "assigns the user as @user" do
+        allow_any_instance_of(User).to receive(:save).and_return(false)
+        put :update, params: { id: user.to_param, user: { name: 'invalid value' } }
+        expect(assigns(:user)).to eq(user)
+      end
 
-        it "does not update the requested user" do
-          expect_any_instance_of(User).not_to receive(:update_attributes)
-          put :update, params: { id: another_user.to_param, user: { name: 'Trung Le' } }
-        end
+      it "re-renders the 'edit' template" do
+        allow_any_instance_of(User).to receive(:save).and_return(false)
+        put :update, params: { id: user.to_param, user: { name: 'invalid value' } }
+        expect(response).to render_template("edit")
+      end
+    end
 
-        it "redirects the user to the homepage" do
-          put :update, params: { id: another_user.to_param, user: valid_attributes }
-          expect(response).to redirect_to(root_url)
-        end
+    context "another user's profile" do
+      let!(:another_user) { create(:user) }
+
+      it "does not update the requested user" do
+        expect_any_instance_of(User).not_to receive(:update_attributes)
+        put :update, params: { id: another_user.to_param, user: { name: 'Trung Le' } }
+      end
+
+      it "redirects the user to the homepage" do
+        put :update, params: { id: another_user.to_param, user: valid_attributes }
+        expect(response).to redirect_to(root_url)
```

Comment:
I would move this part into the example above and also add a test to check that the response body contains the string from the notice message (something like "you're not allowed here").
On their own, these expectations don't provide that much value.

Reply:
There's a few points here that are showing up as new changes, when frankly I haven't touched these lines. Would you recommend removing them all around?